open_spiel/games/gin_rummy.cc

```diff
@@ -57,8 +57,194 @@ std::shared_ptr<const Game> Factory(const GameParameters& params) {
 
 REGISTER_SPIEL_GAME(kGameType, Factory);
 
+bool ObserverHasString(IIGObservationType iig_obs_type) {
+  return !iig_obs_type.perfect_recall;
+}
+
+bool ObserverHasTensor(IIGObservationType iig_obs_type) {
+  return !iig_obs_type.perfect_recall;
+}
+
 }  // namespace
 
+
+class GinRummyObserver : public Observer {
+ public:
+  GinRummyObserver(IIGObservationType iig_obs_type)
+      : Observer(/*has_string=*/ObserverHasString(iig_obs_type),
+                 /*has_tensor=*/ObserverHasTensor(iig_obs_type)),
+        iig_obs_type_(iig_obs_type) {}
+
+
+  void WriteTensor(const State& observed_state, int player,
+                   Allocator* allocator) const override {
+    auto& state = open_spiel::down_cast<const GinRummyState&>(observed_state);
+    SPIEL_CHECK_GE(player, 0);
+    SPIEL_CHECK_LT(player, state.num_players_);
+
+    // TODO(Michal) Should we implement perfect recall?
```

Comment:
If possible, yes. But AFAIK Gin can have various histories lengths, resulting in a lot of "zeros padding" right? Is there a tight upper bound on the length? If it is >>50 it's probably too long.

Reply:
There is no tight upper bound and this made me realize that the max game length needs to be enforced as there are pathological action sequences that can go on indefinitely. This is too long for the perfect recall tensor, but I did implement it for strings following your suggestion. I believe the observer functionality now matches that of dark chess.

Reply:
Great!